Table: Proper handling of json data with dataframes

marefr opened the pull request:
**What this PR does / why we need it**:
When using Raw Document query with Elasticsearch there's a special response from datasource that is used which includes a `type` field with the value `json`. In the table panel there is a transformation for JSON data which up until this fix didn't work at all due to the new data structure we call data frames.

Verified using gdev dashboard/Datasource tests - Elasticsearch v7 and panel "ES Log query"
- renders result in table as before using correct selected columns
- Edit panel the `Table Transform` is `JSON data` and the columns selected in panel model are displayed below. Click the `+` icon gives you a list of all the available fields in the json document.
- Hover a cell in the table allows you to filter for and out using adhoc-filters

**Which issue(s) this PR fixes**:
Fixes #19531

**Special notes for your reviewer**:
Shouldn't be any other datasource using the docs type, at least not in core grafana. Possibly other plugins?

marefr committed:
Table: Proper handling of json data with dataframes

marefr commented on public/app/plugins/panel/table/transformers.ts:
```diff
@@ -209,8 +211,11 @@ transformers['json'] = {
+        const dataFrame = toDataFrame(data[0]);
+        if (dataFrame.meta && dataFrame.fields[0].config.filterable) {
+          tableCol.filterable = true;
+        }
```
All of this feels really messy. Suggestion of improvement very welcome.

ryantxu replied:
kill them all!!!!   I hope we can replace all of this with a new table panel and using 'transformers' rather than the one-off transformers in table panel that are calculated a few ways :(

marefr replied:
Yeah, but the point of this PR is to mainly resolve the critical bug. We have to create a separate issue to do bigger enhancement/migrations eventually.

hugohaggmark reviewed (approved):
LGTM, works as expected! Left a minor comment and couldn't find any small way to refactor JSON transformer. Could probably be good to cover it with tests and then redo it from scratch :)

torkelo commented on public/app/plugins/panel/table/transformers.ts:
```diff
@@ -209,8 +211,11 @@ transformers['json'] = {
-      if (data.length > 0 && data[0].filterable) {
-        tableCol.filterable = true;
+      if (data.length > 0) {
+        const dataFrame = toDataFrame(data[0]);
```
Don't you convert it to DataFrame already in the toLegacy function (convertJSONDocumentDataToDataFrame)

marefr replied:
Can you add some more information, do you want me to enforce the type of `data` argument to be an array of data frames?

torkelo replied:
Sorry, confused, just not sure why we convert it back to DataFrame here? Seems like we convert it to toDataFrame (in runRequest) -> legacy (in MetricsPanelCtrl), then here back to DataFrame. 

Do we need this conversion?

marefr replied:
Hopefully fixed now

ryantxu commented on packages/grafana-data/src/dataframe/processDataFrame.ts:
```diff
@@ -217,6 +244,8 @@ export const guessFieldTypes = (series: DataFrame): DataFrame => {
 
 export const isTableData = (data: any): data is DataFrame => data && data.hasOwnProperty('columns');
 
+export const isJSONDocumentData = (data: any): data is DataFrame => data && data.meta && data.meta.json;
```
do we need this in `@grafana/data`?  can it be isolated to just the table panel?  I don't want to encourage that this gets used anywhere new

marefr replied:
Sure. WIll fix

ryantxu commented on packages/grafana-data/src/dataframe/processDataFrame.ts:
```diff
@@ -241,6 +270,11 @@ export const toDataFrame = (data: any): DataFrame => {
+  if (data.hasOwnProperty('type') && data.type === 'docs') {
+    return convertJSONDocumentDataToDataFrame(data);
+  }
+
```
Is the only difference between this and `convertTimeSeriesToDataFrame` that it adds:
```
    meta: { json: true },
```
If so, do we need the new function?  or just add a note about deprecated/legacy docs mode?

marefr replied:
No, `convertTimeSeriesToDataFrame` expects a time and value besides some other differences. See https://github.com/grafana/grafana/pull/19596/files#diff-0600e6a1b61c8024cc2e5cc95d22c54bR130-R156

marefr committed:
Update public/app/plugins/panel/table/transformers.ts

Co-Authored-By: Hugo Häggmark <hugo.haggmark@grafana.com>

marefr committed:
fix

marefr committed:
isJSONDocumentData as util for table panel

marefr committed:
Handle docs type in toLegacyResponseData

Reverts converting back and forth multiple times between
legacy response data and data frame.

marefr committed:
Merge branch 'master' into 19531_fix

marefr committed:
Merge remote-tracking branch 'origin/master' into 19531_fix

marefr commented on packages/grafana-data/src/dataframe/processDataFrame.ts:
```diff
@@ -288,6 +320,16 @@ export const toLegacyResponseData = (frame: DataFrame): TimeSeries | TableData =
+      target: fields[0].name || frame.name,
+      datapoints: fields[0].values.toArray(),
+      filterable: fields[0].config ? fields[0].config.filterable : undefined,
+      type: 'docs',
```
This isn't really a `TimeSeries`, but not sure if this is okay or not?

torkelo replied:
Think this is ok for now until we make table panel work with DataFrame directly

torkelo reviewed (approved):
Looks good!
